hugegraph-llm/src/operators/build_kg/disambiguate_data.py

```diff
@@ -0,0 +1,239 @@
+Please merge relationships that refer to the same label.
+The datasets contains relationships in the form [LABEL_ID_1, RELATIONSHIP, LABEL_ID_2, PROPERTIES].
+You will also be given a set of LABELS_IDs that are valid.
+Some relationships may use LABELS_IDs that are not in the valid set but refer to a LABEL in the valid set.
+If a relationships refer to a LABELS_IDs in the valid set please change the ID so it matches the valid ID.
+When you have completed your task please give me the valid relationships in the same format. Only return the relationships no other text.
+
+Here is an example
+["Person", "roommate", "Person", {"start": 2021}], ["Person", "owns", "Webpage", {}], ["Person", "roommate", "Person", {"start": 2021}]
+The output you need to provide:
+["Person", "roommate", "Person", {"start": 2021}], ["Person", "owns", "Webpage", {}]
+"""
+
+
+def generate_prompt(data) -> str:
+    return f""" Here is the data:
+{data}
+"""
+
+
+internalRegex = r"\[(.*?)\]"
+
+
+class DisambiguateData:
+    def __init__(self, llm) -> None:
+        self.llm = llm
+
+    def run(self, data: dict) -> dict[str, list[any]]:
+        nodes = sorted(data["nodes"], key=lambda x: x.get("label", ""))
+        relationships = data["relationships"]
+        nodes_schemas = data["nodes_schemas"]
+        relationships_schemas = data["relationships_schemas"]
+        new_nodes = []
+        new_relationships = []
+        new_nodes_schemas = []
+        new_relationships_schemas = []
+
+        node_groups = groupby(nodes, lambda x: x["label"])
+        for group in node_groups:
+            dis_string = ""
+            nodes_in_group = list(group[1])
+            if len(nodes_in_group) == 1:
+                new_nodes.extend(nodes_in_group)
+                continue
+
+            for node in nodes_in_group:
+                dis_string += (
+                    '["'
+                    + node["name"]
+                    + '", "'
+                    + node["label"]
+                    + '", '
+                    + json.dumps(node["properties"])
+                    + "]\n"
+                )
+
+            messages = [
+                {"role": "system", "content": generate_system_message_for_nodes()},
+                {"role": "user", "content": generate_prompt(dis_string)},
+            ]
+            raw_nodes = self.llm.generate(messages)
+            n = re.findall(internalRegex, raw_nodes)
+            new_nodes.extend(nodes_text_to_list_of_dict(n))
+
+        nodes_schemas_data = ""
+        for node_schema in nodes_schemas:
+            nodes_schemas_data += (
+                '["'
+                + node_schema["label"]
+                + '", '
+                + node_schema["primary_key"]
+                + '", '
+                + json.dumps(node_schema["properties"])
+                + "]\n"
+            )
+
+        messages = [
+            {"role": "system", "content": generate_system_message_for_nodes_schemas()},
```

Comment:
If we pass in the schema ourselves, `generate_system_message_for_nodes_schemas()` is not required.

Reply:
yes， I added a boolean value to determine whether the schema is provided by us, to decide whether data cleaning is necessary.